consensus/istanbul/backend/announce.go

```diff
@@ -170,25 +174,40 @@ func (sb *Backend) announceThread() {
 				}
 			}
 
-			updateAnnounceVersionFunc()
-			go sb.generateAndGossipAnnounce(announceVersion)
-
 			// Use this timer to also prune all announce related data structures.
 			if err := sb.pruneAnnounceDataStructures(); err != nil {
 				logger.Warn("Error in pruning announce data structures", "err", err)
 			}
 
-		case <-announceVersionsCheckTicker.C:
-			logger.Trace("Going to check peers' announce version set")
-			go sb.checkPeersAnnounceVersions()
+			sb.startGossipAnnounceTask()
+
+		case <-shareSignedAnnounceVersionTicker.C:
+			// Send all signed announce versions to every peer. Only the entries
+			// that are new to a node will end up being regossiped throughout the
+			// network.
+			allSignedAnnounceVersions, err := sb.getAllSignedAnnounceVersions()
+			if err != nil {
+				logger.Warn("Error getting all signed announce versions", "err", err)
+				break
+			}
+			if err := sb.gossipSignedAnnounceVersionsMsg(allSignedAnnounceVersions); err != nil {
+				logger.Warn("Error gossiping all signed announce versions")
+			}
+
+		case <-sb.generateAndGossipAnnounceCh:
+			if shouldAnnounce {
+				err := sb.generateAndGossipAnnounce()
+				if err != nil {
+					logger.Warn("Error in generating and gossiping announce", "err", err)
+				}
+			}
 
 		case <-sb.updateAnnounceVersionCh:
 			updateAnnounceVersionFunc()
 			sb.updateAnnounceVersionCompleteCh <- struct{}{}
```

Comment:
Something is wrong about this. I don't think it should send to the channel here.

Reply:
I added some comments to make it more clear, as well as a comment in `updateAnnounceVersion` to show that it must be called from a separate goroutine. Does that help clarify what's happening?

Reply:
Looking at it again, it makes more sense 👍